src/resources/tcp_esp32.cc

```diff
@@ -408,26 +384,44 @@ PRIMITIVE(read)  {
       return Smi::from(-1);
     }
 
-    ByteArray* array = allocate_read_buffer(process, p);
+    if (p->tot_len == 0) return Smi::from(-1);  // Nothing to read.
+
+    int offset = socket->read_offset();
+    // Wifi MTU is 1500 bytes, subtract a 20 byte TCP header and we have 1480.
+    const int MAX_SIZE = 496;  // Three nicely-packable byte arrays per 1480 MTU.
+    int tot_len = p->tot_len;
```

Comment:
Move this up before the `p->tot_len == 0` check and maybe give `tot_len` a better local name?